Sources/CvCity.cpp

```diff
@@ -6521,61 +6509,49 @@ void CvCity::recalculateCultureDistances(int iMaxDistance) const
 	bool bHasChanged = (iMaxDistance > 1 ||
 		(iMaxDistance > 0 && GC.getGame().isOption(GAMEOPTION_1_CITY_TILE_FOUNDING)));
 
-	int numLoops = 0;
-	int iNewValue = 0;
 	// as long as there are changes during the last iteration
 	while (bHasChanged)
 	{
 		// reset the has changed variable to note a new loop cycle has begun
 		bHasChanged = false;
 
-		for (int iDX = -iMaxDistance; iDX <= iMaxDistance; iDX++)
+		foreach_(const CvPlot* plotX, plot()->rect(iMaxDistance, iMaxDistance))
 		{
-			for (int iDY = -iMaxDistance; iDY <= iMaxDistance; iDY++)
+			// Some tiles at center should not be calculated with formula (as they already are defined above):
+			if (plotDistance(getX(), getY(), plotX->getX(), plotX->getY()) < 2 &&
+				// Regular realistic culture, with any adjacency to city, and
+				(!GC.getGame().isOption(GAMEOPTION_1_CITY_TILE_FOUNDING) ||
+				// 1-tile-start with cardinal adjacency to city
+				getX() == plotX->getX() || getY() == plotX->getY()))
 			{
-				// Some tiles at center should not be calculated with formula (as they already are defined above):
-				// 1-tile-start with cardinal adjacency to city, and
-				if ((GC.getGame().isOption(GAMEOPTION_1_CITY_TILE_FOUNDING) && (abs(iDX) + abs(iDY) < 2)) ||
-					// regular realistic culture, with any adjacency to city
-					(!GC.getGame().isOption(GAMEOPTION_1_CITY_TILE_FOUNDING) && (abs(iDX) < 2 && abs(iDY) < 2)))
-				{
-					// This is a slightly cursed function.
-					continue;
-				}
-				plotIndex = HASH_RELATIVE_CLOSE_DIST(iDX, iDY);
-				// recalculate the value to determine if it has changed
-				iNewValue = calculateCultureDistance(iDX, iDY, iMaxDistance);
-
-				// if it has changed, save the value and mark that
-				//   ALL values (!) should be recomputed since they
-				//   may depend on this value
-				if (m_aCultureDistances[plotIndex] != iNewValue)
-				{
-					m_aCultureDistances[plotIndex] = iNewValue;
-					bHasChanged = true;
-				}
+				// This is a slightly cursed function.
+				continue;
+			}
+			// recalculate the value to determine if it has changed
+			const int iNewValue = calculateCultureDistance(plotX, iMaxDistance);
+
+			// if it has changed, save the value and mark that
+			//   ALL values (!) should be recomputed since they
+			//   may depend on this value
+			if (m_aCultureDistances[plotX] != iNewValue)
+			{
+				m_aCultureDistances[plotX] = iNewValue;
+				bHasChanged = true;
 			}
-		}
-		numLoops++;
-		if (numLoops > 100)
-		{
-			bHasChanged = false;
```

Comment:
I had issues every now and then with endless loops; easy to accidentially introduce one when fiddling here, and knowing when it's in one vs just taking a slow turn is annoying.

Function really shouldn't be a while(update) loop, but instead something like:
- Spiral out from center
- On current examined tile, look at neighbors to see if there's a shorter distance, and apply to current tile if so
- Attempt to push to neighbors if can be done. If a neighbor was not previously max int and was pushed, add it to a list to come back to later
- Move to next tile in spiral, and continue process.
Once spiral loop is done, do same algo but going down the list, adding entries to bottom of list if tiles get push-updated.

While loop is functional though for now, and it's not obvious how much perf gain there is without profiling.

Reply:
I'll add it back, but you shouldn't need it unless something isn't working right.